gqlalchemy/memgraph.py

```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
+        returns_field = signature[
+            signature.index("(", end_arguments_parantheses) + 1 : signature.index(")", end_arguments_parantheses + 1)
+        ]
+
+        arguments = self._parse_field(arguments_field.strip())
+        returns = self._parse_field(returns_field.strip())
+
+        return arguments, returns
+
+    def _parse_field(self, vars_field: str) -> List[Dict]:
+        """Parse a field of arguments or returns from module signature
+
+        Args:
+            vars_field: signature field inside parantheses
+        """
+        vars = []
+
+        if len(vars_field) == 0:
+            return vars
+
+        list_of_vars = vars_field.split(", ")
+
+        for var in list_of_vars:
+            var_dict = {}
+            sides = var.split(" :: ")
+            var_dict[QM_KEY_TYPE] = sides[1]
+            if " = " in sides[0]:
+                splt = sides[0].split(" = ")
+                var_dict[QM_KEY_NAME] = splt[0]
+                var_dict[QM_KEY_DEFAULT] = splt[1].strip('"')
+            else:
+                var_dict[QM_KEY_NAME] = sides[0]
+
+            vars.append(var_dict)
+
+        return vars
+
+
+class QueryModule:
+    """Class representing a single query module."""
+
+    def __init__(self, module_dict: Dict) -> None:
+        self.name = module_dict[QM_FIELD_NAME]
+        self.is_editable = module_dict[QM_FIELD_IS_EDITABLE]
+        self.is_write = module_dict[QM_FIELD_IS_WRITE]
+        self.path = module_dict[QM_FIELD_PATH]
+        self.signature = module_dict[QM_FIELD_SIGNATURE]
+        self.arguments = module_dict[QM_FIELD_ARGUMENTS]
+        self.returns = module_dict[QM_FIELD_RETURNS]
+
+    def __str__(self) -> str:
+        return self.name
+
+    def set_inputs(self, **kwargs) -> None:
+        """Set values for QueryModule arguments so the module can be called.
+
+        Kwargs:
+            Named arguments in self.arguments.
+
+        Raises:
+            KeyError: Passed an argument not in the self.arguments list.
+        """
+        for kwarg in kwargs.keys():
+            has_arg = False
+            for dict in self.arguments:
+                if dict[QM_KEY_NAME] == kwarg:
+                    dict[QM_KEY_VALUE] = str(kwargs[kwarg])
+                    has_arg = True
+            if not has_arg:
+                raise KeyError(f"{kwarg} is not an argument in this query module.")
+
+    def get_inputs(self) -> str:
+        """return inputs in form "value1, value2, ..." for QueryBuilder call()
+        method.
+
+        Raises:
+            KeyError: Cannot get all values of arguments because one or more is
+            not set.
+        """
+        arguments_str = ""
+        for arg in self.arguments:
+            if QM_KEY_VALUE in arg:
+                val = arg[QM_KEY_VALUE]
+            elif QM_KEY_DEFAULT in arg:
+                val = arg[QM_KEY_DEFAULT]
+            else:
+                raise KeyError(f"{arg[QM_KEY_NAME]} has no value set.")
```

Comment:
Do we have to wrap this error in GQLAchemy errors? Not sure what's the practice, so just asking so that I know too 😄

Reply:
From talking first with Toni and then Ivan and Mrma, KeyError is a standard error and doesn't need to be wrapped